borg/repository.py

```diff
@@ -342,7 +367,26 @@ def _update_index(self, segment, objects, report=None):
                 else:
                     report(msg)
         if self.segments[segment] == 0:
-            self.compact.add(segment)
+            self.compact[segment] += self.io.segment_size(segment)
+
+    def _rebuild_usage(self, segment):
+        """Rebuild compact / usage information for a single segment relative to the current index."""
+        self.compact[segment] = 0
+        self.segments[segment] = 0
+        for tag, key, offset, size in self.io.iter_objects(segment):
+            if tag == TAG_PUT:
+                if self.index.get(key, (-1, -1)) != (segment, offset):
+                    self.compact[segment] += size
+                else:
+                    self.segments[segment] += 1
+            elif tag == TAG_DELETE:
+                s, off = self.index.get(key, (-1, -1))
+                if segment <= s or s == -1:
+                    self.compact[segment] += size
```

Comment:
If the tag_delete is for a key that already doesn't exist then it counts towards free space because it carries no information at all